lib/compress/zstd_fast.c

```diff
@@ -613,75 +633,200 @@ static size_t ZSTD_compressBlock_fast_extDict_generic(
         } else {
-            if ( (matchIndex < dictStartIndex) ||
-                 (MEM_read32(match) != MEM_read32(ip)) ) {
-                assert(stepSize >= 1);
-                ip += ((ip-anchor) >> kSearchStrength) + stepSize;
-                continue;
-            }
-            {   const BYTE* const matchEnd = matchIndex < prefixStartIndex ? dictEnd : iend;
-                const BYTE* const lowMatchPtr = matchIndex < prefixStartIndex ? dictStart : prefixStart;
-                U32 const offset = curr - matchIndex;
-                size_t mLength = ZSTD_count_2segments(ip+4, match+4, iend, matchEnd, prefixStart) + 4;
-                while (((ip>anchor) & (match>lowMatchPtr)) && (ip[-1] == match[-1])) { ip--; match--; mLength++; }   /* catch up */
-                offset_2 = offset_1; offset_1 = offset;  /* update offset history */
-                ZSTD_storeSeq(seqStore, (size_t)(ip-anchor), anchor, iend, OFFSET_TO_OFFBASE(offset), mLength);
-                ip += mLength;
-                anchor = ip;
-        }   }
-
-        if (ip <= ilimit) {
-            /* Fill Table */
-            hashTable[ZSTD_hashPtr(base+curr+2, hlog, mls)] = curr+2;
-            hashTable[ZSTD_hashPtr(ip-2, hlog, mls)] = (U32)(ip-2-base);
-            /* check immediate repcode */
-            while (ip <= ilimit) {
-                U32 const current2 = (U32)(ip-base);
-                U32 const repIndex2 = current2 - offset_2;
-                const BYTE* const repMatch2 = repIndex2 < prefixStartIndex ? dictBase + repIndex2 : base + repIndex2;
-                if ( (((U32)((prefixStartIndex-1) - repIndex2) >= 3) & (offset_2 <= curr - dictStartIndex))  /* intentional overflow */
-                   && (MEM_read32(repMatch2) == MEM_read32(ip)) ) {
-                    const BYTE* const repEnd2 = repIndex2 < prefixStartIndex ? dictEnd : iend;
-                    size_t const repLength2 = ZSTD_count_2segments(ip+4, repMatch2+4, iend, repEnd2, prefixStart) + 4;
-                    { U32 const tmpOffset = offset_2; offset_2 = offset_1; offset_1 = tmpOffset; }  /* swap offset_2 <=> offset_1 */
-                    ZSTD_storeSeq(seqStore, 0 /*litlen*/, anchor, iend, REPCODE1_TO_OFFBASE, repLength2);
-                    hashTable[ZSTD_hashPtr(ip, hlog, mls)] = current2;
-                    ip += repLength2;
-                    anchor = ip;
-                    continue;
-                }
-                break;
-    }   }   }
+            rval = MEM_read32(ip2) ^ 1; /* guaranteed to not match. */
+        }
+
+        /* write back hash table entry */
+        current0 = (U32)(ip0 - base);
+        hashTable[hash0] = current0;
+
+        /* check repcode at ip[2] */
+        if (MEM_read32(ip2) == rval) {
+            ip0 = ip2;
+            match0 = repBase + repIndex;
+            matchEnd = repIndex < prefixStartIndex ? dictEnd : iend;
+            assert((match0 != prefixStart) & (match0 != dictStart));
+            mLength = ip0[-1] == match0[-1];
```

Comment:
Minor : coding style : 
I would recommend to add parenthesis here : 
`mLength = (ip0[-1] == match0[-1]);`

This makes the separation between test and assignment a bit easier to see.

Reply:
Looking back by one single byte is achieved here.
It's unconditional.
Why ? 
An `unlikely()` test could have been possible here, with expected minimal runtime performance impact.

Also : it only looks back by one byte only, not 2+.
Here also, why limited to 1 ?

Reply:
It's designed this way to guarantee that we never accidentally emit a repcode with `litlen == 0` and unintentially change the meaning of the repcode from the one we intended. We could explore changing the structure of this check to be more thorough, but I think we should treat that as out-of-scope of this round of optimizations. I think with the reduced timeframe in which we want to land refactors of `{fast, dfast} * {extDict, dictMatchState}`, we should focus on a fairly direct translation of the `noDict` refactors.